File: app/src/main/java/com/marceljurtz/lifecounter/views/Settings/SettingsActivity.java
package com.marceljurtz.lifecounter.views.Settings;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.app.AlertDialog;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Switch;
import android.widget.TextView;

import com.google.android.material.navigation.NavigationView;
import com.marceljurtz.lifecounter.models.Color;
import com.marceljurtz.lifecounter.enums.MagicColorEnum;
import com.marceljurtz.lifecounter.models.PreferenceManager;
import com.marceljurtz.lifecounter.R;
import com.pes.androidmaterialcolorpickerdialog.ColorPicker;

public class SettingsActivity extends com.marceljurtz.lifecounter.views.Base.View implements ISettingsView {

    //region Controls

    private TextView txtBlack;
    private TextView txtBlue;
    private TextView txtGreen;
    private TextView txtRed;
    private TextView txtWhite;

    private EditText txtLifepoints;
    private EditText txtLongClickPoints;

    private Button cmdSelectBlack;
    private Button cmdSelectBlue;
    private Button cmdSelectGreen;
    private Button cmdSelectRed;
    private Button cmdSelectWhite;

    private Button cmdShowAppIntro;
    private Button cmdReset;

    private Switch chkKeepScreenOn;
    private Switch chkConfirmGameReset;

    private NavigationView navigationView;

    //endregion Controls

    SharedPreferences preferences;
    ISettingsPresenter presenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        preferences = getApplicationContext().getSharedPreferences(PreferenceManager.INSTANCE.getPREFS(), Activity.MODE_PRIVATE);

        initControls();

        disableMenuItem(navigationView, R.id.nav_energy_save_mode);
        disableMenuItem(navigationView, R.id.nav_settings);

        presenter = new SettingsPresenter(this, preferences);
        presenter.onCreate();
    }


    //region Get EditText and color values

    @Override
    public String getLifepoints() {
        return txtLifepoints.getText().toString();
    }

    @Override
    public String getLongClickPoints() {
        return txtLongClickPoints.getText().toString();
    }

    //endregion

    //region Set EditText values

    @Override
    public void setLifepoints(String lifepointText) {
        txtLifepoints.setText(lifepointText);
    }

    @Override
    public void setLongClickPoints(String longClickPointText) {
        txtLongClickPoints.setText(longClickPointText);
    }

    //endregion

    @Override
    public void onBackPressed(){
        presenter.onBackButtonClick();
    }

    @Override
    public void returnToPrevActivity() {
        finish();
    }

    @Override
    public void loadActivity(Class c) {
        Intent intent = new Intent(getApplicationContext(), c);
        startActivity(intent);
    }

    @Override
    public void loadColorPickerDialog(MagicColorEnum color, int r, int g, int b) {
        final ColorPicker cp = new ColorPicker(SettingsActivity.this, r, g, b);
        final MagicColorEnum baseColor = color;

        cp.show();

        Button okColor = (Button) cp.findViewById(R.id.okColorButton);
        okColor.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int newColor = cp.getColor();
                presenter.onColorSelectValueUpdate(new Color(baseColor, newColor));
                cp.dismiss();
            }
        });
    }

    @Override
    public void updateColorButtonValue(Color color) {
        switch(color.getBasecolor()) {
            case BLUE:
                ((GradientDrawable)cmdSelectBlue.getBackground()).setColor(color.getIntValue());
                txtBlue.setText(color.getHexString());
                break;
            case GREEN:
                ((GradientDrawable)cmdSelectGreen.getBackground()).setColor(color.getIntValue());
                txtGreen.setText(color.getHexString());
                break;
            case RED:
                ((GradientDrawable)cmdSelectRed.getBackground()).setColor(color.getIntValue());
                txtRed.setText(color.getHexString());
                break;
            case WHITE:
                ((GradientDrawable)cmdSelectWhite.getBackground()).setColor(color.getIntValue());
                txtWhite.setText(color.getHexString());
                break;
            default:
                ((GradientDrawable)cmdSelectBlack.getBackground()).setColor(color.getIntValue());
                txtBlack.setText(color.getHexString());
                break;
        }
    }

    @Override
    public void loadResetConfirmationDialog() {
        new AlertDialog.Builder(SettingsActivity.this)
                .setMessage(R.string.settings_confirm_reset)
                .setCancelable(false)
                .setPositiveButton(R.string.settings_confirm_reset_true, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        presenter.onResetButtonConfirm();
                    }
                })
                .setNegativeButton(R.string.settings_confirm_reset_false, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        presenter.onResetButtonCancel();
                    }
                })
                .show();
    }

    @Override
    public void setKeepScreenOnCheckbox(boolean checked) {
        chkKeepScreenOn.setChecked(checked);
    }

    @Override
    public void setConfirmGameResetCheckbox(boolean checked) {
        chkConfirmGameReset.setChecked(checked);
    }

    private void initControls() {

        txtBlack = (TextView) findViewById(R.id.txtColorBlack);
        txtBlue = (TextView) findViewById(R.id.txtColorBlue);
        txtGreen = (TextView) findViewById(R.id.txtColorGreen);
        txtRed = (TextView) findViewById(R.id.txtColorRed);
        txtWhite = (TextView) findViewById(R.id.txtColorWhite);

        txtLifepoints = (EditText) findViewById(R.id.txtLiveSelection);
        txtLongClickPoints = (EditText) findViewById(R.id.txtLongClickPoints);

        chkKeepScreenOn = (Switch) findViewById(R.id.chkKeepScreenOn);
        chkKeepScreenOn.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                presenter.onKeepScreenOnCheckboxClick(isChecked);
            }
        });

        chkConfirmGameReset = (Switch) findViewById(R.id.chkConfirmReset);
        chkConfirmGameReset.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean isChecked) {
                presenter.onConfirmGameResetCheckboxClick(isChecked);
            }
        });

        cmdSelectBlack = (Button) findViewById(R.id.cmdSelectBlack);
        cmdSelectBlack.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onColorSelectButtonClick(MagicColorEnum.BLACK);
            }
        });

        cmdSelectBlue = (Button) findViewById(R.id.cmdSelectBlue);
        cmdSelectBlue.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onColorSelectButtonClick(MagicColorEnum.BLUE);
            }
        });

        cmdSelectGreen = (Button) findViewById(R.id.cmdSelectGreen);
        cmdSelectGreen.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onColorSelectButtonClick(MagicColorEnum.GREEN);
            }
        });

        cmdSelectRed = (Button) findViewById(R.id.cmdSelectRed);
        cmdSelectRed.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onColorSelectButtonClick(MagicColorEnum.RED);
            }
        });

        cmdSelectWhite = (Button) findViewById(R.id.cmdSelectWhite);
        cmdSelectWhite.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onColorSelectButtonClick(MagicColorEnum.WHITE);
            }
        });

        cmdReset = (Button) findViewById(R.id.cmdResetSettings);
        cmdReset.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.onResetButtonClick();
            }
        });

        cmdShowAppIntro = (Button)findViewById(R.id.cmdShowIntro);
        cmdShowAppIntro.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                presenter.onShowAppIntroClick();
            }
        });

        navigationView = (NavigationView)findViewById(R.id.navigationViewSettings);
        navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(MenuItem item) {
                switch(item.getItemId()) {
                    case R.id.nav_about:
                        presenter.onMenuEntryAboutTap();
                        break;
                    case R.id.nav_game_2players:
                        presenter.onMenuEntryTwoPlayerTap();
                        break;
                    case R.id.nav_game_4players:
                        presenter.onMenuEntryFourPlayerTap();
                        break;
                    case R.id.nav_dicing:
                        presenter.onMenuEntryDicingTap();
                        break;
                    case R.id.nav_countermanager:
                        presenter.onMenuEntryCounterManagerTap();
                        break;
                    default:
                        break;
                }
                return true;
            }
        });
    }
}

File: app/src/main/java/com/marceljurtz/lifecounter/views/Settings/SettingsActivity.kt
package com.marceljurtz.lifecounter.views.Settings

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.app.AlertDialog
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.CompoundButton
import android.widget.EditText
import android.widget.ImageButton
import android.widget.Switch
import android.widget.TextView

import com.google.android.material.navigation.NavigationView
import com.marceljurtz.lifecounter.models.Color
import com.marceljurtz.lifecounter.enums.MagicColorEnum
import com.marceljurtz.lifecounter.models.PreferenceManager
import com.marceljurtz.lifecounter.R
import com.pes.androidmaterialcolorpickerdialog.ColorPicker

class SettingsActivity : com.marceljurtz.lifecounter.views.Base.View(), ISettingsView {

    //region Controls

    private var txtBlack: TextView? = null
    private var txtBlue: TextView? = null
    private var txtGreen: TextView? = null
    private var txtRed: TextView? = null
    private var txtWhite: TextView? = null

    private var txtLifepoints: EditText? = null
    private var txtLongClickPoints: EditText? = null

    private var cmdSelectBlack: Button? = null
    private var cmdSelectBlue: Button? = null
    private var cmdSelectGreen: Button? = null
    private var cmdSelectRed: Button? = null
    private var cmdSelectWhite: Button? = null

    private var cmdShowAppIntro: Button? = null
    private var cmdReset: Button? = null

    private var chkKeepScreenOn: Switch? = null
    private var chkConfirmGameReset: Switch? = null

    private var navigationView: NavigationView? = null

    //endregion Controls

    internal var preferences: SharedPreferences = getApplicationContext().getSharedPreferences(PreferenceManager.PREFS, Activity.MODE_PRIVATE)
    internal var presenter: ISettingsPresenter = SettingsPresenter( this, preferences)


    //region Get EditText and color values

    //endregion

    //region Set EditText values

    override var lifepoints: String
        get() = txtLifepoints!!.text.toString()
        set(lifepointText) = txtLifepoints!!.setText(lifepointText)

    override var longClickPoints: String
        get() = txtLongClickPoints!!.text.toString()
        set(longClickPointText) = txtLongClickPoints!!.setText(longClickPointText)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)

        preferences = applicationContext.getSharedPreferences(PreferenceManager.PREFS, Activity.MODE_PRIVATE)

        initControls()

        disableMenuItem(navigationView!!, R.id.nav_energy_save_mode)
        disableMenuItem(navigationView!!, R.id.nav_settings)

        presenter = SettingsPresenter(this, preferences)
        presenter.onCreate()
    }

    //endregion

    override fun onBackPressed() {
        presenter.onBackButtonClick()
    }

    override fun returnToPrevActivity() {
        finish()
    }

    override fun loadActivity(c: Class<*>) {
        val intent = Intent(applicationContext, c)
        startActivity(intent)
    }

    override fun loadColorPickerDialog(color: MagicColorEnum, r: Int, g: Int, b: Int) {
        val cp = ColorPicker(this@SettingsActivity, r, g, b)

        cp.show()

        val okColor = cp.findViewById<View>(R.id.okColorButton) as Button
        okColor.setOnClickListener {
            val newColor = cp.color
            presenter.onColorSelectValueUpdate(Color(color, newColor))
            cp.dismiss()
        }
    }

    override fun updateColorButtonValue(color: Color) {
        when (color.basecolor) {
            MagicColorEnum.BLUE -> {
                (cmdSelectBlue!!.background as GradientDrawable).setColor(color.intValue)
                txtBlue!!.text = color.hexString
            }
            MagicColorEnum.GREEN -> {
                (cmdSelectGreen!!.background as GradientDrawable).setColor(color.intValue)
                txtGreen!!.text = color.hexString
            }
            MagicColorEnum.RED -> {
                (cmdSelectRed!!.background as GradientDrawable).setColor(color.intValue)
                txtRed!!.text = color.hexString
            }
            MagicColorEnum.WHITE -> {
                (cmdSelectWhite!!.background as GradientDrawable).setColor(color.intValue)
                txtWhite!!.text = color.hexString
            }
            else -> {
                (cmdSelectBlack!!.background as GradientDrawable).setColor(color.intValue)
                txtBlack!!.text = color.hexString
            }
        }
    }

    override fun loadResetConfirmationDialog() {
        AlertDialog.Builder(this@SettingsActivity)
                .setMessage(R.string.settings_confirm_reset)
                .setCancelable(false)
                .setPositiveButton(R.string.settings_confirm_reset_true) { dialog, id -> presenter.onResetButtonConfirm() }
                .setNegativeButton(R.string.settings_confirm_reset_false) { dialog, which -> presenter.onResetButtonCancel() }
                .show()
    }

    override fun setKeepScreenOnCheckbox(checked: Boolean) {
        chkKeepScreenOn!!.isChecked = checked
    }

    override fun setConfirmGameResetCheckbox(checked: Boolean) {
        chkConfirmGameReset!!.isChecked = checked
    }

    private fun initControls() {

        txtBlack = findViewById<View>(R.id.txtColorBlack) as TextView
        txtBlue = findViewById<View>(R.id.txtColorBlue) as TextView
        txtGreen = findViewById<View>(R.id.txtColorGreen) as TextView
        txtRed = findViewById<View>(R.id.txtColorRed) as TextView
        txtWhite = findViewById<View>(R.id.txtColorWhite) as TextView

        txtLifepoints = findViewById<View>(R.id.txtLiveSelection) as EditText
        txtLongClickPoints = findViewById<View>(R.id.txtLongClickPoints) as EditText

        chkKeepScreenOn = findViewById<View>(R.id.chkKeepScreenOn) as Switch
        chkKeepScreenOn!!.setOnCheckedChangeListener { buttonView, isChecked -> presenter.onKeepScreenOnCheckboxClick(isChecked) }

        chkConfirmGameReset = findViewById<View>(R.id.chkConfirmReset) as Switch
        chkConfirmGameReset!!.setOnCheckedChangeListener { compoundButton, isChecked -> presenter.onConfirmGameResetCheckboxClick(isChecked) }

        cmdSelectBlack = findViewById<View>(R.id.cmdSelectBlack) as Button
        cmdSelectBlack!!.setOnClickListener { presenter.onColorSelectButtonClick(MagicColorEnum.BLACK) }

        cmdSelectBlue = findViewById<View>(R.id.cmdSelectBlue) as Button
        cmdSelectBlue!!.setOnClickListener { presenter.onColorSelectButtonClick(MagicColorEnum.BLUE) }

        cmdSelectGreen = findViewById<View>(R.id.cmdSelectGreen) as Button
        cmdSelectGreen!!.setOnClickListener { presenter.onColorSelectButtonClick(MagicColorEnum.GREEN) }

        cmdSelectRed = findViewById<View>(R.id.cmdSelectRed) as Button
        cmdSelectRed!!.setOnClickListener { presenter.onColorSelectButtonClick(MagicColorEnum.RED) }

        cmdSelectWhite = findViewById<View>(R.id.cmdSelectWhite) as Button
        cmdSelectWhite!!.setOnClickListener { presenter.onColorSelectButtonClick(MagicColorEnum.WHITE) }

        cmdReset = findViewById<View>(R.id.cmdResetSettings) as Button
        cmdReset!!.setOnClickListener { presenter.onResetButtonClick() }

        cmdShowAppIntro = findViewById<View>(R.id.cmdShowIntro) as Button
        cmdShowAppIntro!!.setOnClickListener { presenter.onShowAppIntroClick() }

        navigationView = findViewById<View>(R.id.navigationViewSettings) as NavigationView
        navigationView!!.setNavigationItemSelectedListener { item ->
            when (item.itemId) {
                R.id.nav_about -> presenter.onMenuEntryAboutTap()
                R.id.nav_game_2players -> presenter.onMenuEntryTwoPlayerTap()
                R.id.nav_game_4players -> presenter.onMenuEntryFourPlayerTap()
                R.id.nav_dicing -> presenter.onMenuEntryDicingTap()
                R.id.nav_countermanager -> presenter.onMenuEntryCounterManagerTap()
                else -> {
                }
            }
            true
        }
    }
}